x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,137 @@
+		// return original ACK if withdraw is disabled globally or per channel
+		return ack
+	}
+
+	// unmarshal packet data to obtain the sender and recipient
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		err = sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err.Error())
+	}
+
+	// validate the sender bech32 address from the counterparty chain
+	bech32Prefix := strings.Split(data.Sender, "1")[0]
+	if bech32Prefix == data.Sender {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender: %s", data.Sender).Error(),
+		)
+	}
+
+	senderBz, err := sdk.GetFromBech32(data.Sender, bech32Prefix)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender %s, %s", data.Sender, err.Error()).Error(),
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	if !sender.Equals(recipient) {
+		// return the original success acknowledgement since the address are different
+
+		// NOTE: here we can't know if the recipient is a 'secp256k1' address. If it is,
+		// then the user will need to withdraw the tokens using this same logic.
+		return ack
+	}
+
+	balances := k.bankKeeper.GetAllBalances(ctx, recipient)
+
+	// NOTE: since the balance is 0 we can reject the transaction
+	// We return error since the sender and recipient addresses are the same
+	if balances.IsZero() {
+		logger.Debug(
+			"rejected IBC transfer to 'secp256k1' key address",
+			"sender", data.Sender, "recipient", data.Receiver,
+		)
+
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(types.ErrKeyTypeNotSupported, "receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	// transfer the balance back to the sender address
+
+	srcPort := packet.DestinationPort
+	srcChannel := packet.DestinationChannel
+
+	// swap the sender and recipient
+	sender = recipient
+	recipientStr := data.Sender
+
+	for _, coin := range balances {
+		// withdraw the tokens back to sender
+		if err := k.transferKeeper.SendTransfer(
+			ctx,
+			srcPort,                  // packet destination port is now the source
+			srcChannel,               // packet destination channel is now the source
+			coin,                     // balances + transfer amount
+			sender,                   // transfer recipient is now the sender
+			recipientStr,             // transfer sender is now the recipient
```

Comment:
Shouldn't this recipient string be a custom inputted address by the user? Right now it's sending to an address that is on another chain - I guess this is ok but we need a clear way to IBC transfer from wherever this address is back to a valid Evmos address after?

And this only works with bank balances, not claims records right?

Reply:
only with balances for the moment, correct

Reply:
created a linear issue for it 👍